src/mlpack/methods/svdplusplus/svdplusplus_function_impl.hpp

```diff
@@ -0,0 +1,611 @@
+
+  const arma::mat data = function.Dataset();
+  const arma::sp_mat implicitData = function.ImplicitDataset();
+  const size_t numUsers = function.NumUsers();
+  const size_t numItems = function.NumItems();
+  const double lambda = function.Lambda();
+
+  // Rank of decomposition.
+  const size_t rank = function.Rank();
+
+  // Now iterate!
+  for (size_t i = 1; i != maxIterations; i++, currentFunction++)
+  {
+    // Is this iteration the start of a sequence?
+    if ((currentFunction % numFunctions) == 0)
+    {
+      const size_t epoch = i / numFunctions + 1;
+      Log::Info << "Epoch " << epoch << "; " << "objective "
+          << overallObjective << "." << std::endl;
+
+      // Reset the counter variables.
+      overallObjective = 0;
+      currentFunction = 0;
+    }
+
+    // Indices for accessing the the correct parameter columns.
+    const size_t user = data(0, currentFunction);
+    const size_t item = data(1, currentFunction) + numUsers;
+    const size_t implicitStart = numUsers + numItems;
+
+    // Calculate the squared error in the prediction.
+    const double rating = data(2, currentFunction);
+    const double userBias = parameters(rank, user);
+    const double itemBias = parameters(rank, item);
+
+    // Iterate through each item which the user interacted with to calculate
+    // user vector.
+    arma::vec userVec(rank, arma::fill::zeros);
+    arma::sp_mat::const_iterator it = implicitData.begin_col(user);
+    arma::sp_mat::const_iterator it_end = implicitData.end_col(user);
+    size_t implicitCount = 0;
+    for (; it != it_end; it++)
+    {
+      userVec += parameters.col(implicitStart + it.row()).subvec(0, rank - 1);
+      implicitCount += 1;
+    }
+    if (implicitCount != 0)
+      userVec /= std::sqrt(implicitCount);
+    userVec += parameters.col(user).subvec(0, rank - 1);
+
+    double ratingError = rating - userBias - itemBias -
+        arma::dot(userVec, parameters.col(item).subvec(0, rank - 1));
+
+    // Gradient is non-zero only for the parameter columns corresponding to the
+    // example.
+    parameters.col(user).subvec(0, rank - 1) -= stepSize * 2 * (
+        lambda * parameters.col(user).subvec(0, rank - 1) -
+        ratingError * parameters.col(item).subvec(0, rank - 1));
+    parameters.col(item).subvec(0, rank - 1) -= stepSize * 2 * (
+        lambda * parameters.col(item).subvec(0, rank - 1) -
+        ratingError * userVec);
+    parameters(rank, user) -= stepSize * 2 * (
+        lambda * parameters(rank, user) - ratingError);
+    parameters(rank, item) -= stepSize * 2 * (
+        lambda * parameters(rank, item) - ratingError);
+    // Update item implicit vectors.
+    it = implicitData.begin_col(user);
+    it_end = implicitData.end_col(user);
+    for (; it != it_end; it++)
+    {
+      parameters.col(implicitStart + it.row()).subvec(0, rank - 1) -=
+          stepSize * 2 * (lambda *
+          parameters.col(implicitStart + it.row()).subvec(0, rank - 1) -
```

Comment:
Again `parameters.col(implicitStart + it.row()).subvec(0, rank - 1) / implicitCount`.